[13.0][REF] dms: modify permissions computation

victoralmau opened the pull request:
Related to https://github.com/OCA/dms/pull/102 + Adapt 13.0

- Remove `perm_read` field from `dms.access.group`. It is redundant to the mere existence of the group. Otherwise, users could be able to create, write or unlink records that they cannot read. To avoid confusion, it is removed.
- Add portal group access to mails directory, to make one test pass. This was a false positive before.
- Forbid `dms.access.group` recursiveness.
- Compute inclusive permissions in `dms.access.group`, which combine current group with parent group permissions.
- Rely on domain search of `permission_{create,read,unlink,write}` in `dms.directory` and `dms.file` when checking those accesses to records. These search methods are improved to reflect permission reality and add it to the domain.
- Added a hack to workaround situations where the current context user is superuser, but it's actually checking for other user permissions.
- These checks are added as normal `ir.rule` records.
- To create something in a directory, you need create access in the parent directory.
- For other operations, and only in subdirs, the own dir access is checked.
- Batch operations when checking inherited permissions.
- Remove many custom methods to check access, now that we use ORM's.
- Fix custom cascade dir unlink.
- Use `consteq()` when checking access tokens, for security.
- Remove duplicated security computations from `dms.directory` and `dms.file`. Use proper inheritance.
- Use `ir.rule` for checking locked files.
- Add or fix tests for new permissions model.
- [NEW] Fix prevent error related to mail_message parent not exists in some cases (directory)

Locked: 
- [ ] https://github.com/OCA/dms/pull/106

@Tecnativa TT29847

victoralmau commented:
Please @Yajo can you review it?

yajo commented on dms/models/directory.py:
```diff
@@ -94,6 +112,25 @@ def _default_parent_id(self):
+        comodel_name="dms.category",
+        context="{'dms_category_show_path': True}",
+        string="Category",
+    )
```
These changes do not exist in #102. What happened?

It seems these should come from `abstract.dms.mixin` in v13, including `is_hidden` (which is in v12 but probably not needed here).

Could you check that please?

victoralmau replied:
Sorry, it's not necessary because exists in `abstract.dms.mixin`. Thanks!

yajo commented on dms/models/directory.py:
```diff
@@ -638,7 +647,11 @@ def create(self, vals_list):
                 data = next(iter(parent.sudo().read(["storage_id"])), {})
                 vals["storage_id"] = self._convert_to_write(data).get("storage_id")
-        return super().create(vals_list)
+        # Hack to prevent error related to mail_message parent not exists in some cases
+        ctx = dict(self.env.context).copy()
+        ctx.update({"default_parent_id": False})
+        res = super(DmsDirectory, self.with_context(ctx)).create(vals_list)
```
This code is easy to simplify:

```suggestion
        # Hack to prevent error related to mail_message parent not exists in some cases
        res = super(DmsDirectory, self.with_context(default_parent_id=False)).create(vals_list)
```

I see that you were able to remove this HACK due to the new v13 ORM: https://github.com/OCA/dms/blob/db9cb933ba4f29cc1a610190c32e2ff4bb16b1ce/dms/models/directory.py#L654-L660

Nice!

yajo commented on dms/models/dms_file.py:
```diff
@@ -179,6 +173,7 @@ def check_access_token(self, access_token=False):
         invisible=True,
         ondelete="cascade",
     )
+    storage_id_save_type = fields.Selection(related="storage_id.save_type", store=False)
```
This change does not exist in #102. Is it really needed?

victoralmau replied:
Sorry, it's not necessary because exists in `abstract.dms.mixin`. Thanks!

yajo commented on dms/models/dms_security_mixin.py:
```diff
@@ -2,336 +2,216 @@
-        return [("id", "not in", self.search([])._filter_access_ids("create"))]
+    def _get_permission_domain(self, operator, value, operation):
+        """Abstract logic for searching computed permission fields."""
+        # Tricky one, to know if you want to search
+        # positive or negative access
+        positive = (operator not in NEGATIVE_TERM_OPERATORS) == bool(value)
+        if self.env.uid == SUPERUSER_ID:
+            return TRUE_DOMAIN if positive else FALSE_DOMAIN
```
Another HACK removed from v12: https://github.com/OCA/dms/blob/db9cb933ba4f29cc1a610190c32e2ff4bb16b1ce/dms/models/dms_security_mixin.py#L176-L184

Nice!

yajo commented on dms/security/security.xml:
```diff
@@ -68,34 +81,109 @@
+        <field name="perm_create" eval="1" />
+        <field name="perm_write" eval="0" />
+        <field name="perm_unlink" eval="0" />
+        <field name="domain_force">[('permission_create', '=', True)]</field>
```
And the last HACK from v12 also removed: https://github.com/OCA/dms/blob/db9cb933ba4f29cc1a610190c32e2ff4bb16b1ce/dms/security/security.xml#L119

Nice!

yajo reviewed (changes requested):
Some comments just point out the removed HACK, and that's nice!

However there are some other things to fix.

Also I see that you squashed all commits from #102. No problem on that, but it'd be nice to include in your commit message the contents of all those 3.

Add also `Co-authored-by: Victor Martínez <your-email>` at the end, [to keep co-authorship](https://docs.github.com/en/github/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors), as you totally deserve it.

yajo reviewed (changes requested):
Now the code looks good. However, I found an usability issue.

In v12, when you create a directory assigning no groups (so you're creating an obvious permissions problem), you get this error, telling you you cannot create it:

![image](https://user-images.githubusercontent.com/973709/124261759-1527bd00-db29-11eb-8434-4c536bdea59d.png)

However, now in this v13 runbot, I tested the same and the error is about problems to _read_ the directory:

![screenshot-2021-07-02-11-34-09](https://user-images.githubusercontent.com/973709/124262210-a7c85c00-db29-11eb-9c3a-8776d5ee6f53.png)


All this means that the directory was effectively created, but then, after creating, when trying to read it, there's the access error. I can double-check that by becoming superuser and asserting that it effectively exists:

![screenshot-2021-07-02-11-35-34](https://user-images.githubusercontent.com/973709/124262221-ab5be300-db29-11eb-8bf4-e2e0051cbaec.png)

This is not what should happen. Instead, just like in v12, when creating the record, the creation rules should be checked and should block (or roll back) the record creation, to avoid allowing these senseless conditions.

Commit:
[FIX] dms: respect computed security rules in all contexts

The fix is similar to [v12 creation workaround][1], with some differences:

1. It is applied to both `dms.directory` and `dms.file`, as both need to use this.
2. In v13 Odoo uses now `_filter_access_rules_python()` by default, which is incompatible with DMS' security hack, so we make use of `_filter_access_rules()` always. Some comments explain the HACK.
3. `flush()` before doing the creation access check, so all data is in DB before we issue the `SELECT` needed to check permissions properly.

[1]: https://github.com/OCA/dms/blob/db9cb933ba4f29cc1a610190c32e2ff4bb16b1ce/dms/models/directory.py#L654-L661

Commit:
[REF] dms: modify permissions computation

- Remove `perm_read` field from `dms.access.group`. It is redundant to the mere existence of the group. Otherwise, users could be able to create, write or unlink records that they cannot read. To avoid confusion, it is removed.
- Add portal group access to mails directory, to make one test pass. This was a false positive before.
- Forbid `dms.access.group` recursiveness.
- Compute inclusive permissions in `dms.access.group`, which combine current group with parent group permissions.
- Rely on domain search of `permission_{create,read,unlink,write}` in `dms.directory` and `dms.file` when checking those accesses to records. These search methods are improved to reflect permission reality and add it to the domain.
- Added a hack to workaround situations where the current context user is superuser, but it's actually checking for other user permissions.
- These checks are added as normal `ir.rule` records.
- To create something in a directory, you need create access in the parent directory.
- For other operations, and only in subdirs, the own dir access is checked.
- Batch operations when checking inherited permissions.
- Remove many custom methods to check access, now that we use ORM's.
- Fix custom cascade dir unlink.
- Use `consteq()` when checking access tokens, for security.
- Remove duplicated security computations from `dms.directory` and `dms.file`. Use proper inheritance.
- Use `ir.rule` for checking locked files.
- Add or fix tests for new permissions model.
- Define groups to root folder "Media" (according to the previous expected operation).
- Add the Admin + Demo user group to "Mails" folder (according to the previous expected operation).
- Add sudo() in the creation of the directory to "force" that it can be created because the correct parameter of permission_read can only be correctly defined when the record exists and the complete_group_ids records have been created (in 13.0 it is not necessary).
- Add default group (Admin + Demo user) when creating directive in tests.
- Corrections tests of record_ref
- Fix prevent error related to mail_message parent not exists in some cases (directory)

@Tecnativa TT29847

Co-authored-by: Víctor Martínez <victor.martinez@tecnativa.com>

victoralmau committed:
[FIX] dms: Prevent error (apply sudo()) when recursive directories read from files creation in _compute_path() if some directory not allowed from user

yajo commented:
/ocabot merge major
